contracts/FeralfileArtworkV4.sol

```diff
@@ -1,112 +1,614 @@
 // SPDX-License-Identifier: MIT
-pragma solidity ^0.8.0;
+pragma solidity ^0.8.13;
 
-import "@openzeppelin/contracts/token/ERC721/extensions/ERC721Enumerable.sol";
-import "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
-import "@openzeppelin/contracts/utils/introspection/IERC165.sol";
+import "@openzeppelin/contracts/token/ERC721/ERC721.sol";
 import "@openzeppelin/contracts/utils/Strings.sol";
 
 import "./Authorizable.sol";
 import "./UpdateableOperatorFilterer.sol";
+import "./FeralfileSaleData.sol";
+import "./ECDSASigner.sol";
+
+import "./IFeralfileVault.sol";
 
 contract FeralfileExhibitionV4 is
-    IERC165,
-    ERC721Enumerable,
+    ERC721,
     Authorizable,
-    UpdateableOperatorFilterer
+    UpdateableOperatorFilterer,
+    FeralfileSaleData,
+    ECDSASigner
 {
     using Strings for uint256;
 
+    struct Artwork {
+        uint256 seriesId;
+        uint256 tokenId;
+    }
+
+    struct MintData {
+        uint256 seriesId;
+        uint256 tokenId;
+        address owner;
+    }
+
     // version code of contract
     string public constant codeVersion = "FeralfileExhibitionV4";
 
+    // token base URI
+    string public tokenBaseURI;
+
+    // contract URI
+    string public contractURI;
+
+    // total supply
+    uint256 public totalSupply;
+
     // burnable
-    bool public isBurnable;
+    bool public burnable;
 
     // bridgeable
-    bool public isBridgeable;
+    bool public bridgeable;
 
-    // token base URI
-    string internal _tokenBaseURI;
+    // selling
+    bool private _selling;
 
-    // contract URI
-    string private _contractURI;
+    // mintable
+    bool public mintable = true;
+
+    // cost receiver
+    address public costReceiver;
+
+    // vault contract instance
+    IFeralfileVault public vault;
+
+    // series max supplies
+    mapping(uint256 => uint256) internal _seriesMaxSupplies;
+
+    // series total supplies
+    mapping(uint256 => uint256) internal _seriesTotalSupplies;
+
+    // all artworks
+    mapping(uint256 => Artwork) internal _allArtworks;
+
+    // Mapping from owner to list of owned token IDs
+    mapping(address => uint256[]) private _ownedTokens;
+
+    // Mapping from token ID to index of the owner tokens list
+    mapping(uint256 => uint256) private _ownedTokensIndex;
 
     constructor(
         string memory name_,
         string memory symbol_,
-        bool isBurnable_,
-        bool isBridgeable_
-    ) ERC721(name_, symbol_) {
-        isBurnable = isBurnable_;
-        isBridgeable = isBridgeable_;
+        address signer_,
+        bool burnable_,
+        bool bridgeable_,
+        address vault_,
+        address costReceiver_,
```

Comment:
Small: put the same type params in order. though it wont safe gas in function param, but it's better for reading

Reply:
@lpopo0856 I dont think so, we should put the params based on the meaning like something is similar could put together. In this case, i will bring the signer next to vault and cost receiver.

Reply:
ok!